### PR TITLE
[FIX] ui5Framework: Prevent install of libraries within workspace

### PR DESCRIPTION
**Pull request by matz3**

### CODE REVIEW

**Comment by matz3** [comment]
Ready for review, but tests should be enhanced to cover all the changes.

File: lib/graph/helpers/ui5Framework.js [hunk 1]
```diff
@@ -208,6 +208,58 @@ const utils = {
 			);
 		}
 	},
+	/**
+	 * This logic needs to stay in sync with the dependency definitions for the
+	 * sapui5/distribution-metadata package.
+	 *
+	 * @param {@ui5/project/specifications/Project} project
+	 */
+	async getFrameworkLibraryDependencies(project) {
```
Comment (d3xter666) [inline lib/graph/helpers/ui5Framework.js:217]: I can't find where this is being used

File: lib/graph/helpers/ui5Framework.js [hunk 2]
```diff
@@ -298,7 +349,14 @@ export default {
 
 		log.info(`Using ${frameworkName} version: ${version}`);
 
-		const resolver = new Resolver({cwd: rootProject.getRootPath(), version});
+		let providedLibraryMetadata;
+		if (workspace) {
+			providedLibraryMetadata = await utils.getWorkspaceFrameworkLibraryMetadata({
+				workspace, projectGraph
+			});
+		}
```
[the remark this answers is not in the document]
Comment (matz3) [reply lib/graph/helpers/ui5Framework.js:357]: To simplify the code and behaviour we don't require a version in here anymore. Instead, the AbstractResolve now throws an error before calling `handleLibrary` of the resolver implementation when no version is defined.

File: lib/ui5Framework/AbstractResolver.js [hunk 3]
```diff
@@ -62,7 +59,23 @@ class AbstractResolver {
...
+			promises = {
+				// Use existing metadata if library is provided from outside (e.g. workspace)
+				metadata: Promise.resolve(providedLibraryMetadata),
+				// Provided libraries are already "installed"
+				install: Promise.resolve({
+					pkgPath: providedLibraryMetadata.path
+				})
+			};
+		} else if (!this._version) {
+			throw new Error(`Unable to install library ${libraryName}. No framework version provided.`);
```
Comment (matz3) [inline lib/ui5Framework/AbstractResolver.js:75]: @d3xter666 see here. This is the new version check.

File: lib/graph/helpers/ui5Framework.js [hunk 4]
```diff
@@ -208,6 +208,58 @@ const utils = {
...
+			optionalDependencies = Object.keys(
+				packageInfo.devDependencies || {}
+			).map(($) => $.replace("@openui5/", "")); // @sapui5 dependencies must not be defined in package.json
+		}
+
+		return {dependencies, optionalDependencies};
+	},
+	async getWorkspaceFrameworkLibraryMetadata({workspace, projectGraph}) {
+		const libraryMetadata = Object.create(null);
+		const {projectNameMap} = await workspace._getResolvedModules(); // TODO: private API usage
```
Comment (matz3) [inline lib/graph/helpers/ui5Framework.js:248]: still open

File: lib/graph/helpers/ui5Framework.js [hunk 5]
```diff
@@ -208,6 +208,58 @@ const utils = {
...
+	 * This logic needs to stay in sync with the dependency definitions for the
+	 * sapui5/distribution-metadata package.
+	 *
+	 * @param {@ui5/project/specifications/Project} project
+	 */
+	async getFrameworkLibraryDependencies(project) {
+		let dependencies = [];
+		let optionalDependencies = [];
+
+		// TODO: Add getter for ID
```
Comment (matz3) [inline lib/graph/helpers/ui5Framework.js:221]: still open

**Review by matz3 — COMMENTED** [review]
I'll create a new PR as prerequisite to get rid of the private API usages

File: lib/graph/helpers/ui5Framework.js (hunk 1 above)
Comment (d3xter666) [reply lib/graph/helpers/ui5Framework.js:217]: Ah, sorry, thought it was an external util

File: lib/graph/helpers/ui5Framework.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (d3xter666) [reply lib/graph/helpers/ui5Framework.js:357]: Thanks!

File: lib/graph/helpers/ui5Framework.js [hunk 6]
```diff
@@ -208,6 +208,58 @@ const utils = {
...
+		return {dependencies, optionalDependencies};
+	},
+	async getWorkspaceFrameworkLibraryMetadata({workspace, projectGraph}) {
+		const libraryMetadata = Object.create(null);
+		const {projectNameMap} = await workspace._getResolvedModules(); // TODO: private API usage
+		for (const ui5Module of projectNameMap.values()) {
+			const {project} = await ui5Module.getSpecifications();
+			if (project.isFrameworkProject?.() && !projectGraph.getProject(project.getName())) {
+				const metadata = libraryMetadata[project.getName()] = Object.create(null);
+				metadata.id = project.__id;
```
Comment (d3xter666) [inline lib/graph/helpers/ui5Framework.js:253]: You'd need the getter from line #222 also here

**Comment by d3xter666** [comment]
Apart from TODOs & Tests, the code LGTM.

File: lib/graph/helpers/ui5Framework.js (hunk 6 above)
Comment (matz3) [reply lib/graph/helpers/ui5Framework.js]: Done

File: lib/graph/helpers/ui5Framework.js (hunk 5 above)
Comment (matz3) [reply lib/graph/helpers/ui5Framework.js]: Done

File: lib/graph/helpers/ui5Framework.js (hunk 4 above)
Comment (matz3) [reply lib/graph/helpers/ui5Framework.js]: Done

**Pull request merged**